crates/sui-core/src/narwhal_manager/mod.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use arc_swap::ArcSwap;
+use fastcrypto::bls12381;
+use fastcrypto::traits::KeyPair;
+use futures::future::join_all;
+use mysten_metrics::RegistryService;
+use narwhal_config::{Committee, Parameters, SharedWorkerCache, WorkerId};
+use narwhal_executor::ExecutionState;
+use narwhal_node::{Node, NodeStorage};
+use narwhal_types::ReconfigureNotification;
+use narwhal_worker::TransactionValidator;
+use prometheus::Registry;
+use std::path::PathBuf;
+use std::sync::Arc;
+use sui_types::crypto::NetworkKeyPair;
+use tokio::sync::mpsc::{Receiver, Sender};
+use tokio::task::JoinHandle;
+
+pub struct NarwhalManager {
+    pub join_handle: JoinHandle<()>,
+    pub tx_start: Sender<Arc<Committee>>,
+    pub tx_stop: Sender<()>,
+}
+
+pub struct NarwhalConfiguration<
+    State: ExecutionState + Send + Sync + 'static,
+    TxValidator: TransactionValidator,
+> {
+    pub primary_keypair: bls12381::min_sig::BLS12381KeyPair,
+    pub network_keypair: NetworkKeyPair,
+    pub worker_ids_and_keypairs: Vec<(WorkerId, NetworkKeyPair)>,
+    pub worker_cache: SharedWorkerCache,
+
+    pub storage_base_path: PathBuf,
+    pub parameters: Parameters,
+    pub execution_state: Arc<State>,
+    pub tx_validator: TxValidator,
+
+    pub registry_service: RegistryService,
+}
+
+pub async fn run_narwhal_manager<State, TxValidator>(
+    config: NarwhalConfiguration<State, TxValidator>,
+    mut tr_start: Receiver<Arc<Committee>>,
+    mut tr_stop: Receiver<()>,
+) where
+    State: ExecutionState + Send + Sync + 'static,
+    TxValidator: TransactionValidator,
+{
+    let port = config
+        .parameters
+        .network_admin_server
+        .primary_network_admin_server_port;
+
+    loop {
+        // Copy the config for this iteration of the loop
+        let mut id_keypair_copy = Vec::new();
+        for (id, keypair) in &config.worker_ids_and_keypairs {
+            id_keypair_copy.push((*id, keypair.copy()));
+        }
+
+        // Wait for instruction to start an instance of narwhal
+        let new_committee = match tr_start.recv().await {
+            Some(m) => m,
+            None => break,
+        };
+
+        let new_registry = Registry::new();
+        let _ = config.registry_service.add(new_registry.clone());
```

Comment:
@laura-makdah keep in mind that eventually we'll need to remove the old registry once the node gets shutdown , otherwise we'll keep adding conflicting metrics